File: LightningConstructs/Source/Socket/Methods.cs
namespace Lightning
{
    public partial class Socket
    {
        public Socket()
        {
            
            NetSocket.DualMode = true;
            NetSocket.LingerState = new System.Net.Sockets.LingerOption(true, 30);
        }

        public delegate void SocketConnectedHandler(Socket NewSocket);
        public event SocketConnectedHandler? ServerSocketConnected;
        public event SocketConnectedHandler? ClientSocketConnected;

        private void HookupSocket(System.Net.Sockets.Socket BaseSocket)
        {
            NetSocket = BaseSocket;
            NetSocket.LingerState = new System.Net.Sockets.LingerOption(true, 30);
            IsConnected = true;
        }

        public void Connect(System.Net.IPEndPoint Endpoint)
        {
            if (IsBaseDisposed == true)
            {
                NetSocket = new System.Net.Sockets.Socket(System.Net.Sockets.SocketType.Stream, System.Net.Sockets.ProtocolType.Tcp);
            }
            NetSocket.Connect(Endpoint);
            IsConnected = true;
            
            if (ClientSocketConnected == null)
            {
                Disconnect();
            }
            else
            {
                System.Threading.Thread AsyncThread = new System.Threading.Thread(AsyncThreadMethod => { ClientSocketConnected(this); });
                AsyncThread.Start();
            }
        }
        
        public void Disconnect()
        {
            IsConnected = false;
            NetSocket.Shutdown(System.Net.Sockets.SocketShutdown.Both);
            NetSocket.Disconnect(true);
            NetSocket.Dispose();
            IsBaseDisposed = true;
        }
        
        public void Listen(System.Net.IPEndPoint Endpoint)
        {
            NetSocket.Bind(Endpoint);
            NetSocket.Listen(100);
            IsListening = true;
            var AsyncListener = new System.Threading.Thread(AsyncListenerMethod);
            AsyncListener.Start();
        }

        public void Deafen()
        {
            IsListening = false;
            NetSocket.Disconnect(true);
        }

        public System.Int32 Read(System.Byte[] Buffer, System.Int32 Offset, System.Int32 Length, System.Net.Sockets.SocketFlags Flags)
        {
            lock (ReadLock)
            {
                return NetSocket.Receive(Buffer, Offset, Length, Flags);
            }
            
        }

        public System.Int32 Write(System.Byte[] Buffer, System.Int32 Offset, System.Int32 Length, System.Net.Sockets.SocketFlags Flags)
        {
            lock (WriteLock)
            {
                return NetSocket.Send(Buffer, Offset, Length, Flags);
            }
        }

        private void AsyncListenerMethod()
        {
            do
            {
                if (NetSocket.Poll(0, System.Net.Sockets.SelectMode.SelectRead))
                {
                    var NewSocketBase = NetSocket.Accept();
                
                    var NewSocket = new Socket();
                    NewSocket.HookupSocket(NewSocketBase);
                    if (ServerSocketConnected == null)
                    {
                        NewSocket.Disconnect();
                    }
                    else
                    {
                        System.Threading.Thread AsyncThread = new System.Threading.Thread(AsyncThreadMethod => { ServerSocketConnected(NewSocket); });
                        AsyncThread.Start();
                        //SocketConnected(NewSocket);
                    }
                }
                ListenerGovernor.Limit();
            } while (IsListening);

            
        }
    }
}

File: LightningConstructs/Source/Socket/Methods.vb
Imports System.Runtime.CompilerServices

Public Partial Class Socket
    Implements IDisposable

    Public Sub New()
        NetSocket.DualMode = True
        NetSocket.LingerState = New Net.Sockets.LingerOption(True, 30)
    End Sub

    Public Delegate Sub SocketConnectedHandler(ByVal NewSocket As Socket)

    Public Event SocketConnected As SocketConnectedHandler

    Private Sub HookupSocket(ByVal BaseSocket As Net.Sockets.Socket)
        NetSocket = BaseSocket
        NetSocket.LingerState = New Net.Sockets.LingerOption(True, 30)
        LocalEndPoint = CType(NetSocket.LocalEndPoint, Net.IPEndPoint)
        RemoteEndPoint = CType(NetSocket.RemoteEndPoint, Net.IPEndPoint)
        IsConnected = True
        BaseClientBound = False
    End Sub
    

    
    Private Sub AsyncSocketConnectedMethod(Socket As Socket)
        RaiseEvent SocketConnected(Socket)
    End Sub
    
    Public Sub Connect(ByVal Endpoint As Net.IPEndPoint)
        If IsBaseDisposed = True Then
            NetSocket = New Net.Sockets.Socket(Net.Sockets.SocketType.Stream, Net.Sockets.ProtocolType.Tcp)
        End If
        BaseClientBound = True
        NetSocket.Connect(Endpoint)
        IsConnected = True
        LocalEndPoint = CType(NetSocket.LocalEndPoint, Net.IPEndPoint)
        RemoteEndPoint = CType(NetSocket.RemoteEndPoint, Net.IPEndPoint)
        Dim AsyncThread As Threading.Thread = New Threading.Thread(AddressOf AsyncSocketConnectedMethod)
        AsyncThread.Start(Me)
    End Sub
    
    Public Sub Disconnect()
        IsConnected = False
        NetSocket.Shutdown(Net.Sockets.SocketShutdown.Both)
        NetSocket.Disconnect(True)
        NetSocket.Dispose()
        IsBaseDisposed = True
    End Sub
    
    Public Sub Listen(ByVal Endpoint As Net.IPEndPoint)
        NetSocket.Bind(Endpoint)
        NetSocket.Listen(100)
        IsListening = True
        BaseClientBound = False
        Dim AsyncListener = New Threading.Thread(AddressOf AsyncListenerMethod)
        AsyncListener.Start()
    End Sub
    
    Public Sub Deafen()
        IsListening = False
        NetSocket.Disconnect(True)
    End Sub


    Public Function Read(ByRef Buffer As Byte(), ByVal Offset As Int32, ByVal Length As Int32, ByVal Flags As Net.Sockets.SocketFlags) As Int32
        If NetSocket Is Nothing Then Return 0
        If IsConnected = False Then Return 0
        
        Dim BytesAvailable As Long = NetSocket.Available
        Dim BytesRead As Long
        Dim TotalBytesRead As Long
        Dim TimeoutMS As Long = 1000
        
        ReadTimeoutStopwatch.Restart()
        If BytesAvailable >= Length Then
            Do
                SyncLock ReadLock
                    BytesRead = NetSocket.Receive(Buffer, TotalBytesRead + Offset, Length - TotalBytesRead, Flags)
                End SyncLock
                TotalBytesRead += BytesRead
            Loop Until ReadTimeoutStopwatch.ElapsedMilliseconds >= TimeoutMS Or TotalBytesRead >= Length Or IsConnected = False
        Else
            Do
                SyncLock ReadLock
                    BytesRead = NetSocket.Receive(Buffer, TotalBytesRead + Offset, BytesAvailable - TotalBytesRead, Flags)
                End SyncLock
                TotalBytesRead += BytesRead
            Loop Until ReadTimeoutStopwatch.ElapsedMilliseconds >= TimeoutMS Or TotalBytesRead >= Length Or IsConnected = False
        End If
        ReadTimeoutStopwatch.Stop()
        
        Return TotalBytesRead
    End Function
    
    Public Function Write(ByRef Buffer As Byte(), ByVal Offset As Int32, ByVal Length As Int32, ByVal Flags As Net.Sockets.SocketFlags) As Int32
        If NetSocket Is Nothing Then Return 0
        If IsConnected = False Then Return 0
        Dim TotalBytesWritten As Long
        SyncLock WriteLock
            TotalBytesWritten = NetSocket.Send(Buffer,  Offset, Length, Flags)
        End SyncLock
        Return TotalBytesWritten
    End Function

    Private Sub AsyncListenerMethod()
        Do
            If NetSocket.Poll(0, Net.Sockets.SelectMode.SelectRead) Then
                Dim NewSocketBase = NetSocket.Accept()
                Dim NewSocket = New Socket()
                NewSocket.HookupSocket(NewSocketBase)
                Dim AsyncThread As System.Threading.Thread = New Threading.Thread(AddressOf AsyncSocketConnectedMethod)
                AsyncThread.Start(NewSocket)
            End If
            ListenerGovernor.Limit()
        Loop While IsListening
    End Sub

    <MethodImpl(MethodImplOptions.Synchronized)>
    Public Sub Dispose() Implements IDisposable.Dispose
        NetSocket.Dispose()
    End Sub
End Class
